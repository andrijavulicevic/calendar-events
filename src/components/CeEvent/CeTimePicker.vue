<template>
  <v-menu
    v-model="menu"
    :close-on-content-click="false"
    :nudge-right="40"
    :return-value="time"
    transition="scale-transition"
    offset-y
    full-width
    max-width="290px"
    min-width="290px"
  >
    <template v-slot:activator="{ on }">
      <v-text-field
        v-model="time"
        :label="label"
        prepend-icon="mdi-clock"
        readonly
        v-on="on"
      ></v-text-field>
    </template>
    <v-time-picker
      v-if="menu"
      v-model="time"
      format="24hr"
      no-title
      full-width
      @click:minute="setTime"
    ></v-time-picker>
  </v-menu>
</template>

<script>
export default {
  props: {
    label: {
      type: String,
      required: true
    },
    initialTime: {
      type: String,
      required: true
    }
  },
  data: () => ({
    menu: false,
    time: ""
  }),
  created() {
    this.time = this.initialTime;
  },
  watch: {
    initialTime(newVal) {
      this.time = newVal;
    }
  },
  methods: {
    setTime() {
      this.$emit("timeChanged", this.time);
      this.menu = false;
    }
  }
};
</script>
