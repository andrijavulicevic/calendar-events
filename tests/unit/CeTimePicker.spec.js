import { createLocalVue, mount } from "@vue/test-utils";
import Vuetify from "vuetify";
import CeTimePicker from "../../src/components/CeEvent/CeTimePicker";

const localVue = createLocalVue();
localVue.use(Vuetify);
let vuetify;
let wrapper;
const propsData = {
  label: "Choose time",
  initialTime: "13:00"
};

const createComponent = () =>
  mount(CeTimePicker, { localVue, vuetify, propsData });

describe("CeTimePicker.vue", () => {
  beforeEach(() => {
    vuetify = new Vuetify();
    wrapper = createComponent();
  });

  it("is a Vue instance", () => {
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it("is an CeTimePicker component", () => {
    expect(wrapper.is(CeTimePicker)).toBeTruthy();
  });

  it("should match snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("CeTimePicker props", () => {
    it("has 'label' property", () => {
      expect(wrapper.props("label")).toBeTruthy();
    });

    it("'label' is of type String", () => {
      expect(typeof wrapper.props("label")).toBe("string");
    });

    it("value of 'label' is 'Choose time'", () => {
      expect(wrapper.props("label")).toBe("Choose time");
    });

    it("has 'initialTime' property", () => {
      expect(wrapper.props("initialTime")).toBeTruthy();
    });

    it("'initialTime' is of type String", () => {
      expect(typeof wrapper.props("initialTime")).toBe("string");
    });

    it("vale of 'initialTime' is '13:00'", () => {
      expect(wrapper.props("initialTime")).toBe("13:00");
    });
  });

  describe("CeTimePicker data", () => {
    it("has data", () => {
      expect(CeTimePicker.data).toBeTruthy();
    });

    it("data is of type function", () => {
      expect(typeof CeTimePicker.data).toBe("function");
    });

    it("'menu' is defined and of type Boolean", () => {
      expect(wrapper.vm.menu).toBeDefined();
      expect(typeof wrapper.vm.menu).toBe("boolean");
    });

    it("'menu' has initial value false", () => {
      expect(wrapper.vm.menu).toBe(false);
    });

    it("'time' is defined and of type String", () => {
      expect(wrapper.vm.time).toBeDefined();
      expect(typeof wrapper.vm.time).toBe("string");
    });

    it("'time' has initial value of prop 'initialTime'", () => {
      expect(wrapper.vm.time).toBe(wrapper.props("initialTime"));
    });

    it("'time' changes when prop 'initialTime' is changed", () => {
      wrapper.setProps({ initialTime: "16:00" });
      expect(wrapper.vm.time).toBe("16:00");
    });
  });

  it("event 'timeChanged' emitted", () => {
    wrapper.vm.$emit("timeChanged", "12:00");
    expect(wrapper.emitted().timeChanged).toBeTruthy();
    expect(wrapper.emitted().timeChanged.length).toBe(1);
    expect(wrapper.emitted().timeChanged[0]).toEqual(["12:00"]);
  });

  it("time input value should match 'initialTime' value", () => {
    const timeInput = wrapper.find(".v-input").find("input");
    expect(timeInput.element.value).toBe(wrapper.props("initialTime"));
  });

  it("time input value should change when 'time' is changed", () => {
    const timeInput = wrapper.find(".v-input").find("input");
    expect(timeInput.element.value).toBe("13:00");
    wrapper.vm.time = "16:00";
    expect(timeInput.element.value).toBe("16:00");
  });
});
