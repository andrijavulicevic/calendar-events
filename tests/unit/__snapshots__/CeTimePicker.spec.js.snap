// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CeTimePicker.vue should match snapshot 1`] = `
<div class="v-menu">
  <div class="v-input v-input--is-label-active v-input--is-dirty v-input--is-readonly theme--light v-text-field">
    <div class="v-input__prepend-outer">
      <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-clock theme--light"></i></div>
    </div>
    <div class="v-input__control">
      <div class="v-input__slot">
        <div class="v-text-field__slot"><label for="input-24" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Choose time</label><input id="input-24" readonly="readonly" type="text"></div>
      </div>
      <div class="v-text-field__details">
        <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
      </div>
    </div>
  </div>
  <div role="menu" class="v-menu__content theme--light " style="max-height: auto; min-width: 290px; max-width: 290px; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
</div>
`;
